### PR TITLE
Keyboard shortcuts for post editor: bold, italic, link

### PR DESCRIPTION
**Pull request by ludwiczakpawel**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

It was driving me nuts I couldn't use keyboard shortcuts for post editor so I copy'pasted some code and brought the shortcuts to the editor... It kinds works sooo 🤷‍♂️ :D 

I'm happy to replace it with "more professional" solution though since I'm a JS n00b. And this solution may be completely wrong, I don't know... 

## QA Instructions, Screenshots, Recordings

Go to `/new`, write some text and try `CMD+I`, `CMD+B` and `CMD+K`.

## Added tests?

- [ ] Yes
- [x] No, not needed.
- [ ] I need help with writing tests

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [x] No documentation needed

### CODE REVIEW

**Comment by Link2Twenty** [comment]
We have a keyboard shortcut hook already so it might be better to use that

#10713

You'd probably wanna do something like this
```jsx
import { useKeyboardShortcuts } from '../../shared/components/useKeyboardShortcuts';
import { handleBoldAndItalic, handleBoldAndItalic, handleLink  } from '@utilities/editorKeyboardShortcuts';

useKeyboardShortcuts({
'ctrl+KeyI': handleBoldAndItalic, 
'ctrl+KeyB': handleBoldAndItalic,
'ctrl+KeyK':handleLink ,
}, textAreaRef);
```

Then export your helper functions from editorKeyboardShortcuts

**Comment by ludwiczakpawel** [comment]
@Link2Twenty oh sweeet! :) as I said, I'm a JS moron but I'll try :))) Thanks!

**Comment by Link2Twenty** [comment]
> @Link2Twenty oh sweeet! :) as I said, I'm a JS moron but I'll try :))) Thanks!
> 
> (taking that into consideration, I'm turning this PR into Draft mode)

Would it be helpful if I went through an commented where my bits of code go in, or have you got it? 😊

**Comment by ludwiczakpawel** [comment]
@Link2Twenty absolutely yes!

In fact, we could refactor it in two places if possible because we also have these functions on initializeCommentsPage.js.erb too

File: app/javascript/utilities/editorKeyboardShortcuts.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+const KEY_CODE_B = 66;
+const KEY_CODE_I = 73;
```
Comment (Link2Twenty) [inline app/javascript/utilities/editorKeyboardShortcuts.js:2]: This line can go

File: app/javascript/utilities/editorKeyboardShortcuts.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+const KEY_CODE_K = 75;
```
Comment (Link2Twenty) [inline app/javascript/utilities/editorKeyboardShortcuts.js:3]: This line can go

File: app/javascript/utilities/editorKeyboardShortcuts.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+export function handleKeyDown(e) {
```
Comment (Link2Twenty) [inline app/javascript/utilities/editorKeyboardShortcuts.js:5]: This function can go

File: app/javascript/utilities/editorKeyboardShortcuts.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        e.preventDefault();
+        handleLink(e);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+function handleBoldAndItalic(event) {
```
Comment (Link2Twenty) [inline app/javascript/utilities/editorKeyboardShortcuts.js:26]: this function will be exported

File: app/javascript/utilities/editorKeyboardShortcuts.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  );
+
+  const selectionStartWithOffset = selectionStart + surroundingStr.length;
+  textArea.setSelectionRange(
+    selectionStartWithOffset,
+    selectionStartWithOffset + selection.length,
+  );
+}
+
+function handleLink(event) {
```
Comment (Link2Twenty) [inline app/javascript/utilities/editorKeyboardShortcuts.js:48]: this function will be exported

**Comment by Link2Twenty** [comment]
So [execCommand](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand) is obsolete meaning we probably shouldn't use it. [setRangeText](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setRangeText) is fine to use but does not put changes into the undo pool. I can continue to look into a solution. Here is modification to your helper functions to allow toggling and to not use `execCommand`.

```js
export function handleBold(event) {
  event.preventDefault();

  _toggleSurroundText(event.target, '**');
}

export function handleItalic(event) {
  event.preventDefault();

  _toggleSurroundText(event.target, '__');
}

export function handleLink(event) {
  event.preventDefault();
  const textArea = event.target;

  const selection = textArea.value.substring(
    textArea.selectionStart,
    textArea.selectionEnd,
  );
  const selectionStart = textArea.selectionStart;
  const selectionEnd = textArea.selectionEnd;

  textArea.setRangeText(`[${selection}](url)`, selectionStart, selectionEnd);

  // start position + length of selection + [](
  const startOffset = selectionStart + selection.length + 3;

  // start offset + 'url'.length
  const endOffset = startOffset + 3;

  textArea.setSelectionRange(startOffset, endOffset);
}

function _toggleSurroundText(textArea, surroundingStr) {
  // text to modify
  const selection = textArea.value.substring(
    textArea.selectionStart,
    textArea.selectionEnd,
  );

  // surrounding area for checks
  const expandedSelection = textArea.value.substring(
    textArea.selectionStart - surroundingStr.length,
    textArea.selectionEnd + surroundingStr.length,
  );

  // current selection position
  let selectionStart = textArea.selectionStart;
  let selectionEnd = selectionStart + selection.length

  // position and string once transformed
  let newString = `${surroundingStr}${selection}${surroundingStr}`;
  let newStart = selectionStart + surroundingStr.length;
  let newEnd = selectionEnd + surroundingStr.length;

  // if it's already been set to bold/italics toggle back.
  if (expandedSelection === newString) {
    // selection position to include existing tag
    selectionStart = selectionStart - surroundingStr.length;
    selectionEnd = selectionEnd + surroundingStr.length;

    // position and string once transformed
    newString = selection;
    newStart = selectionStart;
    newEnd = newStart + selection.length;
  }

  textArea.setRangeText(newString, selectionStart, selectionEnd);
  textArea.setSelectionRange(newStart, newEnd);
}

```

**Comment by citizen428** [comment]
Hi @ludwiczakpawel, were you thinking about picking this up again? If not, maybe @Link2Twenty would be nice enough to address this in a new PR?

**Comment by ludwiczakpawel** [comment]
@citizen428 I'd happy to hand it off.

**Comment by thomasbnt** [comment]
Oooh yay, **interesting** topic 👀

**Comment by citizen428** [comment]
@thomasbnt Would you be interested in taking this over and finishing it? 😃

**Comment by Link2Twenty** [comment]
Someone over on [forem.dev](https://forem.dev/) mentioned they'd made an [external rich text editor](https://forem.dev/saravanan10393/option-to-upload-the-attachments-on-behalf-of-user-using-api-key-4g5c) that will produce forem valid markdown. 

I think moving over to real rich text would solve this issue but also make a better editing experience in general. It's too big a project for me to commit to though 😅

**Links:**
- <https://deveditor.dev/>
- <https://github.com/saravanan10393/dev-editor/tree/master>
- <https://docs.slatejs.org/>

**Comment by thomasbnt** [comment]
> @thomasbnt Would you be interested in taking this over and finishing it?

Hi! I'm [in procedure install](https://docs.forem.com/installation/linux/) (On Arch) for the first time, and I don't know if I'm going to get it all right. Maybe I could take this PR later, but it's not sure so if someone can do it, especially since it looks pretty complicated

**Comment by citizen428** [comment]
@Link2Twenty Since you commented here previously, any chance you want to take over this PR and finish it? 😃 No worries if you don't have time for that, I just thought I'd ask.

**Comment by Link2Twenty** [comment]
@citizen428 with the version 3 editor (Rich Text/WYSIWYG) on the roadmap for this year is it worth waiting for that? I would presume that would have all the keyboard shortcuts built in.

**Comment by citizen428** [comment]
Good point @Link2Twenty. I guess I was mostly looking for something quick that can bridge that gap (i.e. just incorporating some of the changes you suggested here) but I understand if that's not super appealing.

**Comment by Link2Twenty** [comment]
I just discovered something interesting.

I made a note over on #13234 about how we could use `contenteditable` `div`s rather than `textarea`s, here's the [demo](https://jsfiddle.net/link2twenty/8hmrb7d4/) again, because they auto size change. What I didn't realise at the time was they also take ctrl+b and ctrl+i by default.

cc @aitchiss

**Comment by aitchiss** [comment]
Hey @Link2Twenty I don't think we want to move our text areas over to `contenteditable` for this particular PR. As the editor and the backend rely on markdown, we need the text content of the area to change to e.g. `**bold**` on `ctrl + b`, so even in chrome where this looks like it's working, it doesn't meet our needs unfortunately 🙁 

I think the shortcuts in this PR would still be useful, like @citizen428 mentioned, bridging the gap until we have more functionality in the editor. If you'd like to take it on, let us know, but of course no worries if not!

**Comment by citizen428** [comment]
Hi @Link2Twenty 👋 Do you think you'll work on this PR? If not it's probably time to close it for now and revisit this topic later.

**Comment by Link2Twenty** [comment]
I don't think I'll have time to work on this I'm afraid.

**Pull request closed without merging**